server/src/infra/graphql/mutation.rs

```diff
@@ -163,6 +172,52 @@ fn unpack_attributes(
     })
 }
 
+/// Consolidates caller supplied user fields and attributes into a list of attributes.
+///
+/// A number of user fields are internally represented as attributes, but are still also
+/// available as fields on user objects. This function consolidates these fields and the
+/// given attributes into a resulting attribute list. If a value is supplied for both a
+/// field and the corresponding attribute, the attribute will take precedence.
+fn consolidate_attributes(
+    attributes: Vec<AttributeValue>,
+    first_name: Option<String>,
+    last_name: Option<String>,
+    avatar: Option<String>,
+) -> Vec<AttributeValue> {
+    // Prepare map of the client provided attributes
+    let mut provided_attributes: BTreeMap<AttributeName, AttributeValue> = attributes
+        .into_iter()
+        .map(|x| {
+            (
+                x.name.clone().into(),
+                AttributeValue {
+                    name: x.name.clone().to_lowercase(),
+                    value: x.value,
+                },
+            )
+        })
+        .collect::<BTreeMap<_, _>>();
+    // Prepare list of fallback attribute values
+    let field_attrs: Vec<(&str, Option<String>)> = vec![
+        ("first_name", first_name),
+        ("last_name", last_name),
+        ("avatar", avatar),
+    ];
+    for a in field_attrs.iter() {
```

Comment:
I think with ` into_iter` we can move the values out of the array, and then we avoid another copy of the attribute value (the `val.to_owned()`)

Reply:
Hm, not sure I can do that. `val` comes from the `if let Some(val) ...`, and it won't let me move that out of the `AttributeValue` anyway.

Reply:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=a41b6de40cb81ccddd0695a9db02cf35

Reply:
Note: the field is an array, not a slice
Into_iter
The pattern matching doesn't borrow

Reply:
Ah, right, didn't know I could unpack it in the for loop.

Reply:
Should be good now.